qiskit/transpiler/layout.py

```diff
@@ -374,4 +374,36 @@ class TranspileLayout:
 
     initial_layout: Layout
     input_qubit_mapping: Dict[Qubit, int]
+    output_qubit_mapping: Dict[Qubit, int]
     final_layout: Optional[Layout] = None
+
+    @property
+    def num_qubits(self) -> int:
+        """Number of qubits in the represented Layout."""
+        return len(self.input_qubit_mapping)
+
+    def get_initial_permutation(self) -> list[int]:
+        """Get initial permutation from initial layout."""
+        inverted_input_mapping = {i: q for q, i in self.input_qubit_mapping.items()}
+        input_qubits = tuple(inverted_input_mapping[i] for i in range(self.num_qubits))
+        return [self.initial_layout[q] for q in input_qubits]
+
+    def get_final_permutation(self) -> list[int]:
+        """Get final permutation from final layout."""
+        inverted_output_mapping = {i: q for q, i in self.output_qubit_mapping.items()}
+        output_qubits = tuple(inverted_output_mapping[i] for i in range(self.num_qubits))
+        final_layout = self.final_layout or Layout(self.output_qubit_mapping)
```

Comment:
An example of why setting `final_layout` to `None` instead of the trivial layout adds complexity @jakelishman @jlapeyre. 

I would personally advocate for enforcing it (i.e. `final_layout: Layout` instead of `final_layout: Optional[Layout] = None`).